ADR 007: Semantic Relations Decorator Pattern (#71)
pyx-industries/julee: docs/ADRs/007-semantic-relations.md

# ADR 007: Semantic Relations Decorator Pattern

## Status

Draft

## Date

2026-01-07

## Context

Julee solutions consist of multiple bounded contexts, each with their own entities. These entities often have meaningful relationships across bounded context boundaries:

- A `Story` in the HCD bounded context relates to a `UseCase` in core
- An `Accelerator` provides a view onto a `BoundedContext`
- A solution's `CustomerSegment` entity specializes the framework's `Persona`

These relationships are **semantic** - they express architectural intent and domain meaning, not just runtime data flow. They differ from:

1. **Import dependencies** - which are compile-time coupling
2. **Foreign key relationships** - which are instance-level data references
3. **Event subscriptions** - which are runtime behavior coupling

Without explicit semantic relationships:

- Documentation cannot show cross-BC navigation (e.g., "what relates to UseCase?")
- Architectural validation cannot verify that entities properly declare their intent
- Introspection tools cannot discover how bounded contexts conceptually relate

## Decision

> Entity types SHALL declare semantic relationships to other entity types in other bounded contexts, **where those semantic relationships cannot otherwise be inferred by AST/code analysis**. The `@semantic_relation` decorator attaches relationship metadata to classes, discoverable via introspection.

Author: the purpose of semantic relations is to improve documentation. Imagine you are looking at sphinx docs for a class, and it has all the usual stuff (derived from parameter/return types, inheritance, etc). The sphinx directives can add other links too if we need them - is imported/used by, uses/imports, etc - using AST/code analysis. So there's no need to declare them, if we need it we can figure it out. But the AST/code analysis can not infer semantic intent, that's why we need to make those assertions explicitly.

> Entity types SHALL declare semantic relationships to other entity types in other bounded contexts, **where those semantic relationships cannot otherwise be inferred by AST/code analysis**. The `@semantic_relation` decorator attaches relationship metadata to classes, discoverable via introspection.

Author: (if we want to have "see also: these things which are semantically related in this way" type cross-references in the documentation)

```

This enables infrastructure (documentation, introspection) to navigate the relationship graph in both directions without hardcoding entity-to-entity mappings.

### Relation Labels for Documentation

Each relation type has forward and inverse labels for human-readable documentation:

| RelationType | Forward Label | Inverse Label |
|--------------|---------------|---------------|
| `PROJECTS` | "Projects" | "Projected by" |
| `ENABLES` | "Enables" | "Enabled by" |
| `REFERENCES` | "References" | "Referenced by" |
| `PART_OF` | "Part of" | "Contains" |
| `CONTAINS` | "Contains" | "Part of" |

### String References for Circular Imports

When entities have circular import relationships, use string paths:

```python
# In story.py - cannot import App directly due to circular import
@semantic_relation("julee.hcd.entities.app.App", RelationType.PART_OF)
class Story(BaseModel):
    ...
```

The string is resolved to an actual type at decoration time via dynamic import.

### Doctrine Constraints

Semantic relations are subject to doctrine:

1. **source_type and target_type MUST be BaseModel or Enum subclasses** - only doctrine-valid entity types can participate in semantic relations

> 2. **Types MUST be actual types, not strings** - the SemanticRelation entity stores resolved types, not string references

Reviewer: ```suggestion
2. **Type string references MUST resolve to actual types** - the SemanticRelation entity stores resolved types, not the string references
```
